// react-progressbar/index.js
/**
 * @jsx React.DOM
 */
'use strict';

var React = require('react');

var component = React.createClass({
  propTypes: {
    onProgressChange: React.PropTypes.func,
  },
    
  render: function() {

    var containerStyle = {
      height: '10px'
    };

    var progressStyle = {
      backgroundColor: this.props.color || '#0BD318',
      width: (this.props.completed || 0) + '%',
      transition: "width 200ms",
      height: '10px'
    };

    return (
      <div ref='parent' className="progressbar-container" style={containerStyle} onClick = {this._handleClick}>
        <div className="progressbar-progress" style={progressStyle}>
        </div>
      </div>
    );
  },

    _handleClick: function(evt, args) {
	console.error('click');
	var div = this.refs.parent;
	if (!div) {
	    return;
	}
	
	var rect = div.getDOMNode().getBoundingClientRect();
	var left = rect.left; 
	var width = rect.width;
	var click_left = evt.pageX;

	var percent = (click_left - left) / width * 100;
	if (this.props.onProgressChange) {
	    this.props.onProgressChange(percent);
	}
	console.error(percent);
    }

    
});

module.exports = component;
